File: sort/quick-sort.js
const { swap, left, right } = require('./util');

/**
 * Takes -> O ( k )
 * 
 * @param {*} arr 
 * @param {*} leftIndex 
 * @param {*} rightIndex 
 */
exports.partition = (arr, leftIndex, rightIndex) => {

    let pivot = arr[rightIndex];
    let i = leftIndex = -1;

    for (let j = leftIndex; j < rightIndex; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr, i, j);
        }
    }
    i++;
    swap(arr, i, rightIndex);
    return i;
}

exports.quicksort = (arr, leftIndex, rightIndex) => {
    if (leftIndex < rightIndex) {
        let middle = this.partition(arr, leftIndex, rightIndex);
        this.quicksort(arr, leftIndex, middle - 1);
        this.quicksort(arr, middle + 1, rightIndex);
    }
}

exports.sort = (arr) => {
    this.quicksort(arr, 0, arr.length - 1);
}

File: src/main/java/com/github/danimaniarqsoft/sorting/QuickSort.java
package com.github.danimaniarqsoft.sorting;

import static com.github.danimaniarqsoft.utils.CommonUtils.swap;

public class QuickSort implements Sort {

    @Override
    public void sort(int[] array) {
        quickSort(array, 0, array.length - 1);
    }

    private void quickSort(int[] array, int leftIndex, int rightIndex) {

        if (leftIndex < rightIndex) {
            int pivotIndex = partition(array, leftIndex, rightIndex);
            quickSort(array, leftIndex, pivotIndex - 1);
            quickSort(array, pivotIndex + 1, rightIndex);
        }
    }

    private int partition(int[] array, int leftIndex, int rightIndex) {
        int pivot = array[rightIndex];
        int minIndex = leftIndex - 1;

        for (int i = leftIndex; i < rightIndex; i++) {
            if (array[i] < pivot) {
                swap(array, i, ++minIndex);
            }
        }

        swap(array, ++minIndex, rightIndex);
        return minIndex;
    }

}
